### ISSUE TITLE
Angular CT providers are not added to the TestingModule

### ISSUE THREAD

**Issue by NicBright** [cypress-io/cypress#23427]
### Current behavior

The bug is obvious:

https://github.com/cypress-io/cypress/blob/d201b37b3d6b1e37a15a8d21d853acca47bfc666/npm/angular/src/mount.ts#L143-L157

The providers that are meant to be added to the TestingModule are instead only added in a call to `TestBed.overrideComponent()`. This way, it is not possible to alter dependencies of singleton services aka services that are `{ providedIn: 'root' }` (or any custom module).

### Desired behavior

Passed providers **must** be passed to `TestBed.configureTestingModule()`

### Test code to reproduce

Irrelevant. See source code snippet above.

### Cypress Version

10.6.0

### Node version

irrelevant

### Operating System

irrelevant

**Comment by NicBright** [comment]
Until this bug is fixed, for those of you that stumble upon this. There's a workaround:

Create and add a dedicated "YourXyzTestingModule" to your `imports`. Within the definition of `YourXyzTestingModule` you can add providers which will be used as if you've added them to the `providers` config directly.

**Comment by mike-plummer** [comment]
Hi @NicBright , we originally had the implementation you're asking for but we reverted it [based on this issue](https://github.com/jordanpowell88/angular-ct/issues/37). The component under test may need a custom implementation of a given provider but other uses rendered in that test may need another, and separating the providers in this way rather than forcing them to the TestBed level makes that a bit easier. For situations where you want a singleton provider your identified workaround (nice job finding that, btw) would be a solution. Of course, that may not be obvious to many users so this may be useful to note in our angular documentation

**Comment by jordanpowell88** [comment]
@NicBright I am unable to duplicate the issue you are describing. I created an example of what you are describing [here](https://github.com/cypress-io/cypress-component-testing-apps/pull/7). Like @mike-plummer mentioned we are purposely overriding the component providers at the component level

**Comment by NicBright** [comment]
@jordanpowell88
Refering to your example [here](https://github.com/cypress-io/cypress-component-testing-apps/pull/7)
Of course you can still override the LoginService to inject a double inside an Angular CT this way **for your component under test.** However, you can neither inject a double for a child component, nor can you inject a double to be used as a dependency for any other singleton service. For example: how would you inject a double for HttpClient as it is required by LoginService?

**Comment by jordanpowell88** [comment]
@NicBright I am still not able to reproduce the issue you are describing. I just created a the following test and it is working the way you are describing it would not:

`@Component({
  template: `
    <h1>Parent Component</h1>
    <app-child-providers></app-child-providers>
  `
})
export class ParentProvidersComponent {}`

`@Injectable()
export class ChildProvidersService {
  constructor(private readonly http: HttpClient) {}

  getData(): Observable<string> {
    return this.http.get<string>('https://myfakeapiurl.com/api/data')
  }
}

@Component({
  selector: 'app-child-providers',
  template: `<button (click)="handleClick()">{{ message }}</button>`
})
export class ChildProvidersComponent {
  message = 'default message'

  constructor(private readonly service: ChildProvidersService) {}

  handleClick(): void {
    this.service.getData().pipe(
      take(1)
    ).subscribe(value => {
      this.message = value
    })
  }
}`

`it('can make test doubles for child components', () => {
    cy.mount(ParentProvidersComponent, {
      declarations: [ChildProvidersComponent],
      imports: [HttpClientModule],
      providers: [
        {
          provide: ChildProvidersService,
          useValue: {
            getData() {
              return of('test')
            }
          } as ChildProvidersService
        }
      ]
    })`

**Comment by NicBright** [comment]
@jordanpowell88 

The following example reproduces the two issues:
```typescript
@Component({
    template: `
        <h1>Parent Component</h1>
        <app-child-providers></app-child-providers>
    `,
})
export class ParentProvidersComponent {}

@Injectable({
    providedIn: 'root',
})
export class ChildProvidersService {
    constructor(private readonly http: HttpClient) {}

    getData(): Observable<string> {
        return this.http.get<string>('https://myfakeapiurl.com/api/data');
    }
}

@Component({
    selector: 'app-child-providers',
    template: `<button (click)="handleClick()">{{ message }}</button>`,
})
// eslint-disable-next-line rxjs-angular/prefer-takeuntil
export class ChildProvidersComponent {
    message = 'default message';

    constructor(private readonly service: ChildProvidersService) {}

    handleClick(): void {
        this.service
            .getData()
            .pipe(take(1))
            // eslint-disable-next-line rxjs-angular/prefer-takeuntil
            .subscribe((value) => {
                this.message = value;
            });
    }
}

@Component({
    selector: 'app-child-providers',
    template: `<button (click)="handleClick()">{{ message }}</button>`,
    providers: [ChildProvidersService], // such a provider cannot be replaced with a double!
})
// eslint-disable-next-line rxjs-angular/prefer-takeuntil
export class AnotherChildProvidersComponent {
    message = 'default message';

    constructor(private readonly service: ChildProvidersService) {}

    handleClick(): void {
        this.service
            .getData()
            .pipe(take(1))
            // eslint-disable-next-line rxjs-angular/prefer-takeuntil
            .subscribe((value) => {
                this.message = value;
            });
    }
}

describe('Cypress issue 23427', () => {
    it('This works - it seems that providers added to a component are also available for child components', () => {
        cy.mount(ParentProvidersComponent, {
            declarations: [ChildProvidersComponent],
            imports: [],
            providers: [
                {
                    provide: ChildProvidersService,
                    useValue: {
                        getData() {
                            return of('test');
                        },
                    } as ChildProvidersService,
                },
            ],
        });
        cy.contains('default message').click();
        cy.contains('test');
    });

    it('However they are NOT available for child components when they have dedicated @Component({providers}) configured (this test will fail)', () => {
        cy.mount(ParentProvidersComponent, {
            declarations: [AnotherChildProvidersComponent],
            imports: [HttpClientTestingModule],
            providers: [
                {
                    provide: ChildProvidersService,
                    useValue: {
                        getData() {
                            return of('test');
                        },
                    } as ChildProvidersService,
                },
            ],
        });
        cy.contains('default message').click();
        cy.contains('test');
    });

    it('You also cannot provide a double for a dependency of a singleton service (this test will fail)', () => {
        cy.mount(ParentProvidersComponent, {
            declarations: [ChildProvidersComponent],
            imports: [],
            providers: [
                { // try to provide a double for ChildProvidersService's dependency to HttpClient:
                    provide: HttpClient,
                    useValue: {
                        get() {
                            return of('test');
                        },
                    },
                },
            ],
        });
        cy.contains('default message').click();
        cy.contains('test');
    });
});
```

Result:
![image](https://user-images.githubusercontent.com/1225659/186867094-aed482ca-e5c3-434e-b331-01b5b73292e2.png)

**Comment by jordanpowell88** [comment]
Hey Nic. Thanks again for providing the thorough example of your problem.

The best way to accomplish what you are wanting to do at the moment is to use something like the following:

`TestBed.overrideProvider(ChildProvidersService, {
      useValue: {
        getMessage() {
          return of('test')
        }
      } as ChildProvidersService
    })
    `
    
You still have access to the same Static `TestBed` we are using for mount so you should be able to overrideProviders that way. We may do some additional investigation into adding first party override support via #23425 but this should be sufficient work around for the time being

**Comment by ZachJW34** [comment]
Going to reopen, as `providers` makes sense to map to the TestModule rather than the component since all of the other `TestModuleMetadata` properties are currently being mapped to the TestModule. We should add an API for overriding at the component level which could be relevant for normal components and Standalone components.

**Comment by jordanpowell88** [comment]
This is solved in #24394
